internal/state/mailbox_search.go

```diff
@@ -4,425 +4,581 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/ProtonMail/gluon/internal/db"
+	"github.com/bradenaw/juniper/parallel"
+	"github.com/bradenaw/juniper/xslices"
 	"net/mail"
+	"runtime"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/ProtonMail/gluon/imap"
 	"github.com/ProtonMail/gluon/internal/contexts"
-	"github.com/ProtonMail/gluon/internal/db"
 	"github.com/ProtonMail/gluon/internal/db/ent"
 	"github.com/ProtonMail/gluon/internal/parser/proto"
 	"github.com/ProtonMail/gluon/rfc822"
-	"github.com/bradenaw/juniper/xslices"
 	"golang.org/x/text/encoding"
 )
 
+var totalActiveSearchRequests int32
+
 func (m *Mailbox) Search(ctx context.Context, keys []*proto.SearchKey, decoder *encoding.Decoder) ([]uint32, error) {
-	snapMessages := m.snap.getAllMessages()
+	var mapFn func(snapMsgWithSeq) uint32
 
-	messages, err := doSearch(ctx, m, snapMessages, keys, decoder)
+	if contexts.IsUID(ctx) {
+		mapFn = func(s snapMsgWithSeq) uint32 {
+			return uint32(s.UID)
+		}
+	} else {
+		mapFn = func(s snapMsgWithSeq) uint32 {
+			return uint32(s.Seq)
+		}
+	}
+
+	op, err := buildSearchOpListWithKeys(m, keys, decoder)
 	if err != nil {
 		return nil, err
 	}
 
-	return xslices.Map(messages, func(msg snapMsgWithSeq) uint32 {
-		if contexts.IsUID(ctx) {
-			return uint32(msg.UID)
+	msgCount := m.snap.len()
+
+	result := make([]uint32, msgCount)
+
+	activeSearchRequests := atomic.AddInt32(&totalActiveSearchRequests, 1)
+	defer atomic.AddInt32(&totalActiveSearchRequests, -1)
+
+	parallelism := runtime.NumCPU() / int(activeSearchRequests)
```

Comment:
we do the same thing for fetch -- wondering two things:
- should we merge the two counters? i'd expect fetching and searching to take each other into account wrt parallelism
- should we abstract this logic into an object to more tidily handle this? the object could return you the amount of parallelism you're allowed to use at any given time

Reply:
```go
type limiter struct {
  // hold the number of ongoing jobs
}

func (l *limiter) do(fn func(limit int)) {
  // determine how much parallelism can be applied right now
  
  // increase ongoing job counter
  
  fn(limit) // call the callback with the amount allowed
  
  // decrease ongoing job counter
}
```

Reply:
I initially thought about limiting the total operations for both fetch and search. But I found that since most of these queries are IO bound, they don't really stress the CPU a lot. I never go 100% cpu usage during these benchmarks. 

I think it's safe to have ncpu fetches and search in parallel.

I would leave this query decision to each command. E.g: with search it makes sense to parallelize all the time since you need to scan the whole inbox.

Reply:
hmm okay 👍